src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -83,22 +83,94 @@ public Dictionary(int capacity, IEqualityComparer<TKey>? comparer)
         public Dictionary(IDictionary<TKey, TValue> dictionary) : this(dictionary, null) { }
 
         public Dictionary(IDictionary<TKey, TValue> dictionary, IEqualityComparer<TKey>? comparer) :
-            this(dictionary != null ? dictionary.Count : 0, comparer)
+            this(0, comparer)
         {
             if (dictionary == null)
             {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.dictionary);
             }
 
+            int count = dictionary.Count;
+            if (count < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity);
+            }
+            if (count == 0)
+            {
+                // Nothing to copy, all done
+                return;
+            }
+
             // It is likely that the passed-in dictionary is Dictionary<TKey,TValue>. When this is the case,
             // avoid the enumerator allocation and overhead by looping through the entries array directly.
             // We only do this when dictionary is Dictionary<TKey,TValue> and not a subclass, to maintain
             // back-compat with subclasses that may have overridden the enumerator behavior.
             if (dictionary.GetType() == typeof(Dictionary<TKey, TValue>))
             {
                 Dictionary<TKey, TValue> d = (Dictionary<TKey, TValue>)dictionary;
-                int count = d._count;
+
+                if (d._comparer == _comparer)
+                {
+                    // Comparers are the same, can copy dictionary wholesale
+                    if (d._freeCount == 0)
+                    {
+                        // No free spots, dictionary is compact so copy arrays directly
+                        int[]? buckets = d._buckets;
+                        if (buckets is not null)
+                        {
+                            _buckets = new int[buckets.Length];
+                            buckets.CopyTo(_buckets.AsSpan());
+                        }
+
+                        Entry[]? oldEntries = d._entries;
+                        if (oldEntries is not null)
+                        {
+                            _entries = new Dictionary<TKey, TValue>.Entry[oldEntries.Length];
+                            oldEntries.CopyTo(_entries.AsSpan());
+                        }
+    #if TARGET_64BIT
+                        _fastModMultiplier = d._fastModMultiplier;
+    #endif
+                        _count = d._count;
+                        _freeList = d._freeList;
+                        _freeCount = d._freeCount;
+                        return;
+                    }
+                    else
+                    {
+                        // Are free spots, so use TrimExcess code to conditionally copy the entries
```

Comment:
Is this optimization for this case worth the extra code? Dictionaries with removed entries are pretty rare.

Reply:
I might have misunderstood what you meant then?

> Is it going to regress performance if somebody is using this method to compact the dictionary?

Reply:
I think you understand it correctly. The current implementation of the change is going to improve performance for some cases and regress performance for other cases. I am poking at whether it would be better to avoid the regressions, at the cost of giving up some performance for the best case.

Reply:
Now uses trimmed path for oversized dictionaries and direct copy for right sized dictionaries